cgen: V auto functions is not working in Sumtype variants declaration

### Describe the bug

cgen Error

### Reproduction Steps

```v
module main

struct Chacabum {
	a int
}

type SumType = Chacabum | int | string


fn (v SumType) cast[T](val T) string {

	mut res := 'lala'

	$for variant_value in v.variants {
		if v is variant_value {
			// println(v) // ok
			// println('v: ${v.str()}') // ok
			// println('v: ${v}') // not ok <-----------------------------------------------------------------
			// return v // ok
			// res = v.str() // ok
			// res = 'v: ${v.str()}' // ok
			res = 'v: ${v}' // not ok <-----------------------------------------------------------------
		}
	}

	return res

}

fn main() {
	assert SumType(1).cast[int](1) == 'v: 1'
}
```

### Expected Behavior

`assert SumType(1).cast[int](1) == 'v: 1'` pass

### Current Behavior

```sh
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/call_methods_at_comptime2.01J91SAWTKJPEVTEFH0X1S8VT4.tmp.c:13382: error: cannot convert 'struct main__Chacabum' to 'unsigned char *'
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

### Possible Solution

make `res = 'v: ${v}'` works as `res = 'v: ${v.str()}'`?

### Additional Information/Context

_No response_

### V version

V 0.4.8 aa3ab3f

### Environment details (OS name and version, etc.)

Ubuntu
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/str_intp.v
+++ b/vlib/v/gen/c/str_intp.v
@@ -191,7 +191,7 @@ fn (mut g Gen) str_val(node ast.StringInterLiteral, i int, fmts []u8) {
 	} else if fmt == `s` || typ.has_flag(.variadic) {
 		mut exp_typ := typ
 		if expr is ast.Ident {
-			if g.comptime.is_comptime_var(expr) {
+			if g.comptime.get_ct_type_var(expr) == .smartcast {
 				exp_typ = g.comptime.get_comptime_var_type(expr)
 			} else if expr.obj is ast.Var {
 				if expr.obj.smartcasts.len > 0 {

--- a/vlib/v/gen/c/str_intp.v
+++ b/vlib/v/gen/c/str_intp.v
@@ -191,7 +191,9 @@ fn (mut g Gen) str_val(node ast.StringInterLiteral, i int, fmts []u8) {
 	} else if fmt == `s` || typ.has_flag(.variadic) {
 		mut exp_typ := typ
 		if expr is ast.Ident {
-			if expr.obj is ast.Var {
+			if g.comptime.is_comptime_var(expr) {
+				exp_typ = g.comptime.get_comptime_var_type(expr)
+			} else if expr.obj is ast.Var {
 				if expr.obj.smartcasts.len > 0 {
 					exp_typ = g.unwrap_generic(expr.obj.smartcasts.last())
 					cast_sym := g.table.sym(exp_typ)

--- a/vlib/v/tests/comptime/comptime_variant_interp_test.v
+++ b/vlib/v/tests/comptime/comptime_variant_interp_test.v
@@ -1,0 +1,23 @@
+module main
+
+struct Foo {
+	a int
+}
+
+type SumType = Foo | int | string
+
+fn (v SumType) cast[T](val T) string {
+	mut res := ''
+	$for variant_value in v.variants {
+		if v is variant_value {
+			println('v: ${v}')
+			res = 'v: ${v}'
+			println(res)
+		}
+	}
+	return res
+}
+
+fn test_main() {
+	assert SumType(1).cast[int](1) == 'v: 1'
+}

